src/porepy/numerics/ad/ad_system.py

```diff
@@ -0,0 +1,1969 @@
+"""Contains the AD system manager, managing variables and equations for a system modelled
+using the AD framework.
+
+"""
+
+from __future__ import annotations
+
+import itertools
+from enum import Enum, EnumMeta
+from typing import Any, Callable, Literal, Optional, Union, overload
+
+import numpy as np
+import scipy.sparse as sps
+
+import porepy as pp
+
+from . import _ad_utils
+from .operators import MixedDimensionalVariable, Operator, Variable
+
+__all__ = ["SystemManager"]
+
+GridLike = Union[pp.Grid, pp.MortarGrid]
+"""A union type representing a domain either by a grid or mortar grids."""
+
+VarLike = Union[
+    str,
+    list[str],
```

Comment:
We have previously decided to enforce lists, i.e. passing [single_variable], in AD module. I suggest doing so here as well.